### PR TITLE
SPI: Add support for STM32F4{01,07,12,29}

### PR DESCRIPTION
**Pull request by psycowithespn**
I based this off of spi.rs in stm32f103xx-hal. Verified on an F407 with a device and F429 by logic analyzer / loopback.

The SPI4/5/6 block should be left commented until https://github.com/adamgreig/stm32-rs/pull/83 is fully released (next version) and this crate updates to the new version.

### CODE REVIEW

File: src/spi.rs [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+            impl<PINS> Spi<$SPIX, PINS> {
+                pub fn $spiX(
+                    spi: $SPIX,
+                    pins: PINS,
+                    mode: Mode,
+                    freq: Hertz,
+                    clocks: Clocks
+                ) -> Self
+                where PINS: Pins<$SPIX> {
+
```
Comment (therealprof) [inline src/spi.rs]:
```suggestion
```

File: src/spi.rs [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+    }
+}
+
+hal! {
+    SPI1: (spi1, apb2enr, spi1en, pclk2),
+    SPI2: (spi2, apb1enr, spi2en, pclk1),
+    SPI3: (spi3, apb1enr, spi3en, pclk1),
+}
+
+// TODO(psycowithespn) Enable all of these on the next version of stm32-rs.
```
Comment (therealprof) [inline src/spi.rs]: Please remove this block for now. I don't want random todos hanging around in the code.

**Comment by therealprof** [comment]
Thanks for the PR, LGTM except for the todo block and the one nit...

**Comment by psycowithespn** [comment]
Applied both requested changes. 

I can squash it if you'd like.

**Pull request merged**